[12.0][MIG] purchase & purchase_stock

MiquelRForgeFlow opened the pull request:
Migration scripts for `purchase` and `purchase_stock` modules.




--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

MiquelRForgeFlow committed:
[MIG] purchase

MiquelRForgeFlow committed:
[MIG] purchase_stock

cuongnmtm commented:
I used your migration and got this issue:
```
  File "/opt/odoo/code/projects/OpenUpgrade/odoo/models.py", line 3739, in _load_records_write
    self.write(values)
  File "/opt/odoo/code/projects/OpenUpgrade/addons/mrp/models/stock_warehouse.py", line 237, in write
    return super(StockWarehouse, self).write(vals)
  File "/opt/odoo/code/projects/OpenUpgrade/addons/stock/models/stock_warehouse.py", line 167, in write
    warehouse._create_or_update_global_routes_rules()
  File "/opt/odoo/code/projects/OpenUpgrade/addons/stock/models/stock_warehouse.py", line 247, in _create_or_update_global_routes_rules
    self[rule_field] = self.env['stock.rule'].create(values)
  File "<decorator-gen-3>", line 2, in create
  File "/opt/odoo/code/projects/OpenUpgrade/odoo/api.py", line 451, in _model_create_multi
    return create(self, [arg])
  File "/opt/odoo/code/projects/OpenUpgrade/odoo/models.py", line 3523, in create
    records = self._create(data_list)
  File "/opt/odoo/code/projects/OpenUpgrade/odoo/models.py", line 3623, in _create
    cr.execute(query, params)
  File "/opt/odoo/code/projects/OpenUpgrade/odoo/sql_db.py", line 148, in wrapper
    return f(self, *args, **kwargs)
  File "/opt/odoo/code/projects/OpenUpgrade/odoo/sql_db.py", line 225, in execute
    res = self._obj.execute(query, params)
odoo.tools.convert.ParseError: "null value in column "picking_type_id" violates not-null constraint
DETAIL:  Failing row contains (7, 1, WH: Stock → Pre-Production (MTO), 20, 1, 1, f, 2019-02-18 05:03:22.545127, pull, 2019-02-18 05:03:22.545127, null, propagate, null, 20, 15, null, 0, 1, t, make_to_order, null, 1, null, null, manual).
" while parsing /opt/odoo/code/projects/OpenUpgrade/addons/purchase_stock/data/purchase_stock_data.xml:15, near
<record id="stock.warehouse0" model="stock.warehouse">
            <field name="buy_to_resupply" eval="True"/>
        </record>
```

nkhuyit commented:
I got an error while I was trying to migrate model purchase. Can anyone help me to handle it, please?
Traceback (most recent call last):
  File "/opt/odoo/code/projects/OpenUpgrade/odoo/http.py", line 1469, in dispatch
    odoo.registry(db).check_signaling()
  File "/opt/odoo/code/projects/OpenUpgrade/odoo/__init__.py", line 81, in registry
    return modules.registry.Registry(database_name)
  File "/opt/odoo/code/projects/OpenUpgrade/odoo/modules/registry.py", line 62, in __new__
    return cls.new(db_name)
  File "/opt/odoo/code/projects/OpenUpgrade/odoo/modules/registry.py", line 86, in new
    odoo.modules.load_modules(registry._db, force_demo, status, update_module)
  File "/opt/odoo/code/projects/OpenUpgrade/odoo/modules/loading.py", line 476, in load_modules
    force, status, report, loaded_modules, update_module, models_to_check, upg_registry)
  File "/opt/odoo/code/projects/OpenUpgrade/odoo/modules/loading.py", line 360, in load_marked_modules
    upg_registry=upg_registry,
  File "/opt/odoo/code/projects/OpenUpgrade/odoo/modules/loading.py", line 222, in load_module_graph
    registry.init_models(cr, model_names, {'module': package.name})
  File "/opt/odoo/code/projects/OpenUpgrade/odoo/modules/registry.py", line 304, in init_models
    models[0].recompute()
  File "/opt/odoo/code/projects/OpenUpgrade/odoo/models.py", line 5264, in recompute
    vals = {n: rec[n] for n in ns}
  File "/opt/odoo/code/projects/OpenUpgrade/odoo/models.py", line 5264, in <dictcomp>
    vals = {n: rec[n] for n in ns}
  File "/opt/odoo/code/projects/OpenUpgrade/odoo/models.py", line 5094, in __getitem__
    return self._fields[key].__get__(self, type(self))
  File "/opt/odoo/code/projects/OpenUpgrade/odoo/fields.py", line 1006, in __get__
    self.determine_value(record)
  File "/opt/odoo/code/projects/OpenUpgrade/odoo/fields.py", line 1093, in determine_value
    self.compute_value(recs)
  File "/opt/odoo/code/projects/OpenUpgrade/odoo/fields.py", line 1073, in compute_value
    self._compute_value(records)
  File "/opt/odoo/code/projects/OpenUpgrade/odoo/fields.py", line 1064, in _compute_value
    getattr(records, self.compute)()
  File "/opt/odoo/code/projects/OpenUpgrade/addons/purchase/models/purchase.py", line 645, in _compute_product_uom_qty
    line.product_uom_qty = line.product_uom._compute_quantity(line.product_qty, line.product_id.uom_id)
  File "/opt/odoo/code/projects/OpenUpgrade/addons/uom/models/uom_uom.py", line 152, in _compute_quantity
    raise UserError(_('The unit of measure %s defined on the order line doesn\'t belong to the same category than the unit of measure %s defined on the product. Please correct the unit of measure defined on the order line or on the product, they should belong to the same category.') % (self.name, to_unit.name))
odoo.exceptions.UserError: ("The unit of measure 0.95 liter/box defined on the order line doesn't belong to the same category than the unit of measure Liter(s) defined on the product. Please correct the unit of measure defined on the order line or on the product, they should belong to the same category.", '')

pedroslopez commented:
Checking this it seems fine, but after migrating there is an issue with the domains applied on the purchase order actions.

On v12 I see `purchase_rfq` no longer has a domain applied, so the old one stays. `purchase_form_action` has an different domain which _is_ loaded.  The old domain for `purchase_rfq` and new domain for `purchase_form_action` conflict in a way, since neither includes orders that are "to approve" so there's no way of seeing them (this is how we found out).

The domain for the `purchase_rfq` action should be removed on migration.

To me, that action shouldn't have had the domain removed entirely, but that's besides the point and off topic

MiquelRForgeFlow commented:
Good catch!! thanks!

@pedrobaeza should we make the `openupdate_records` detect this kind of changes?

rvalyi commented on addons/purchase/migrations/12.0.1.2/post-migration.py:
```diff
@@ -0,0 +1,38 @@
+        UPDATE ir_act_window iaw
+        SET domain = NULL
+        FROM ir_model_data imd
+        WHERE AND imd.name = 'purchase_rfq' AND imd.res_id = iaw.id
```
bad syntax: `WHERE AND`
I guess you meant `WHERE imd.name = 'purchase_rfq' AND imd.res_id = iaw.id` ?

MiquelRForgeFlow replied:
yes, thanks, it was a typo

pedrobaeza committed:
[FIX] purchase: Refine migration

- Log querys
- Delete mail templates translations
- product_uom_qty computation rewritten.
  * Field added with openupgradelib method.
  * Performance optimization via SQL.
  * You can't change without user knowledge UoM of already confirmed POs, as
    stock.move UoM won't match.
- purchase.route_warehouse0_buy XML-ID renamed on purchase, not on
  purchase_stock, as being new module, migration scripts are not executed.
- Nothing to do on purchase_stock

pedrobaeza commented:
I have fixed following things:

- Log querys
- Delete mail templates translations
- product_uom_qty computation rewritten.
  * Field added with openupgradelib method.
  * Performance optimization via SQL.
  * You can't change without user knowledge UoM of already confirmed POs, as stock.move UoM won't match.
- purchase.route_warehouse0_buy XML-ID renamed on purchase, not on
  purchase_stock, as being new module, migration scripts are not executed.
- Nothing to do on purchase_stock.
- Simplified domain removal.

pedrobaeza commented:
/ocabot merge

OthmaneGhandi commented:
(Thanks in advance) @mreficent & @pedrobaeza I got this error while trying to confirm purchase order ("location_dest_id" violates not-null constraint) I verified and it seems all things are correct, I don't know the origin of it

`2019-07-28 10:02:02,128 17960 ERROR db_v12_17_06_2019 odoo.sql_db: bad query: b'INSERT INTO "stock_move" ("id", "create_uid", "create_date", "write_uid", "write_da$
ERROR: null value in column "location_dest_id" violates not-null constraint
DETAIL:  Failing row contains (4672, PI-PO00028, 2019-07-28 10:02:01.795525, null, 1, null, 200, make_to_stock, null, null, 1, 26, 55, 10, 1, null, draft, null, null, $

2019-07-28 10:02:02,131 17960 INFO db_v12_17_06_2019 werkzeug: 46.114.33.84 - - [28/Jul/2019 10:02:02] "POST /web/dataset/call_button HTTP/1.1" 200 - 533 0.171 0.1$
2019-07-28 10:02:28,875 17960 INFO db_v12_17_06_2019 werkzeug: 46.114.33.84 - - [28/Jul/2019 10:02:28] "POST /longpolling/poll HTTP/1.1" 200 - 9 0.004 50.008
2019-07-28 10:03:12,944 17960 INFO db_v12_17_06_2019 werkzeug: 46.114.33.84 - - [28/Jul/2019 10:03:12] "POST /web/dataset/call_kw/purchase.order/default_`

pedrobaeza commented:
Check your picking type default locations. This doesn't seem directly related to the migration.
